include/deal.II/base/timer_tree.h

```diff
@@ -0,0 +1,423 @@
+    AssertThrow(id == ids[0], ExcMessage("Invalid ID specified."));
+
+    std::vector<std::string> remaining_id = erase_first(ids);
+
+    bool found = false;
+    if (remaining_id.size() > 0)
+      {
+        for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+          {
+            if ((*it)->id == remaining_id[0])
+              {
+                (*it)->insert(remaining_id, sub_tree, new_name);
+                found = true;
+              }
+          }
+      }
+
+    if (found == false)
+      {
+        AssertThrow(
+          remaining_id.size() == 0,
+          ExcMessage(
+            "Subtree can not be inserted since the specified ID does not exist in this tree."));
+
+        std::shared_ptr<TimerTree> new_tree(new TimerTree());
+        new_tree->copy_from(sub_tree);
+        if (!new_name.empty())
+          new_tree->id = new_name;
+
+        // Make sure that the new tree does not already exist
+        for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+          {
+            AssertThrow(
+              new_tree->id != (*it)->id,
+              ExcMessage(
+                "Subtree can not be inserted since the tree already contains a subtree with the same ID."));
+          }
+
+        sub_trees.push_back(new_tree);
+      }
+  }
+
+  /**
+   * Prints wall time of all items of a tree without an analysis of
+   * the relative share of the children.
+   */
+  void
+  print_plain(ConditionalOStream const &pcout) const
+  {
+    unsigned int const length = get_length();
+
+    pcout << std::endl;
+
+    do_print_plain(pcout, 0, length);
+  }
+
+  /**
+   * This is the actual function of interest of this class, i.e., an
+   * analysis of wall times with a hierarchical formatting of results.
+   * Relative wall times are printed for all children of a sub-tree if
+   * a wall time has been set for the parent of these children. In this
+   * case, an additional item `other` is created in order to give insights
+   * to which extent the code has been covered with timers and to which
+   * extend time is spent in other code paths that are currently not
+   * covered by timers. The parameter @p level specifies the level for which
+   * to print results, where a value of 0 corresponds to the high level module.
+   */
+  void
+  print_level(ConditionalOStream const &pcout, unsigned int const level) const
+  {
+    unsigned int const length = get_length();
+
+    pcout << std::endl;
+
+    do_print_level(pcout, level, 0, length);
+  }
+
+private:
+  /*
+   * Copy function.
+   */
+  void
+  copy_from(std::shared_ptr<TimerTree const> const other)
+  {
+    *this = *other;
+  }
+
+  /*
+   * This function erases the first entry of the vector.
+   */
+  std::vector<std::string>
+  erase_first(std::vector<std::string> const &in) const
+  {
+    AssertThrow(in.size() > 0, ExcMessage("empty name."));
+
+    std::vector<std::string> out(in);
+    out.erase(out.begin());
```

Comment:
If you do this operation frequently, you should look at using `std::deque` instead of `std::vector` since that allows for `pop_front()`.

Reply:
I think this is something one could also change later.